File: java/examples/src/main/java/org/pactman/nonprofitcheckplus/examples/cases/Ex16NotFound.java
package org.pactman.nonprofitcheckplus.examples.cases;

import org.pactman.nonprofitcheckplus.Ein;
import org.pactman.nonprofitcheckplus.devtools.FixtureEins;
import org.pactman.nonprofitcheckplus.examples.support.Example;
import org.pactman.nonprofitcheckplus.examples.support.ExampleContext;
import org.pactman.nonprofitcheckplus.examples.support.ExampleFailedException;
import org.pactman.nonprofitcheckplus.examples.support.Output;
import org.pactman.nonprofitcheckplus.exceptions.PactmanNotFoundException;

/**
 * EX-16 — A well-formed EIN with no record.
 *
 * <p>"Shaped like an EIN" and "an EIN the IRS has a record for" are different
 * questions. The first is answered locally; only the second costs a request, and
 * a miss is an ordinary outcome rather than a fault.
 */
public final class Ex16NotFound implements Example {

    @Override
    public String id() {
        return "ex-16";
    }

    @Override
    public String title() {
        return "A valid EIN the API has no record for";
    }

    @Override
    public void run(String[] args) {
        String ein = FixtureEins.NO_RECORD;

        Output.heading("Locally, this EIN is fine");
        Output.field("Ein.isValid", Ein.isValid(ein));
        Output.field("Ein.normalize", Ein.normalize(ein));

        try (ExampleContext context = ExampleContext.withFixtures()) {
            Output.heading("The API has no record for it");

            try {
                context.client().nonprofits().check(ein);
                throw new ExampleFailedException("Expected the lookup to report no record.");
            } catch (PactmanNotFoundException failure) {
                Output.field("exception", failure.getClass().getSimpleName());
                Output.field("category", failure.category());
                Output.field("origin", failure.origin());
                Output.field("status", failure.status());
                Output.field("message", failure.getMessage());
                Output.field("requestId", failure.requestId());
                Output.field("attempts", failure.attempts());
            }
        }

        Output.note("A single check with no record is an HTTP 404, and the SDK raises it as\n"
                + "PactmanNotFoundException — which is never retried, because the answer will\n"
                + "not change on a second attempt.\n\n"
                + "Bulk behaves differently on purpose: a batch where some EINs matched is a\n"
                + "200, with the misses listed in getNotFoundEins(). See ex-19.\n\n"
                + "No record is not evidence that an organization does not exist. It is the\n"
                + "absence of a record in the extracts behind this API.");
    }
}

File: ruby/examples/ex_16_not_found.rb
# frozen_string_literal: true

# EX-16 — EIN not found, and application-level failures.
#
# A well-formed EIN with no matching record is a normal outcome, not a bug. The
# single endpoint answers HTTP 404, which the SDK raises as `NotFoundError` — a
# subclass of `ApiError`, so a handler can rescue the specific case or the general
# one.
#
# The envelope's own `code`, `message` and `errors` survive onto the error, and
# none of the diagnostics contain the API key.
#
# Run:  PACTMAN_API_KEY=... bundle exec ruby examples/ex_16_not_found.rb

require_relative "lib/fixture_api"
require_relative "lib/print"

NCP = Pactman::NonprofitCheckPlus
include Print

FixtureApi.with_fixture_api do |client|
  no_record = FixtureApi::EINS[:no_record]

  heading "Single check for #{no_record} — well formed, no record"

  begin
    result = client.nonprofits.check(no_record)
    puts "  Unexpectedly succeeded: #{result.nonprofit&.organization_name}"
  rescue NCP::NotFoundError => e
    # Stable identity: class, category, and origin. Never parse `message`.
    field "class", e.class.name
    field "category", e.category.inspect
    field "origin", e.origin.inspect
    field "is an ApiError", e.is_a?(NCP::ApiError)
    field "is an NCP::Error", e.is_a?(NCP::Error)
    field "matches the not_found category", e.category == NCP::ErrorCategory::NOT_FOUND

    heading "  Response detail carried on the error"
    field "status", e.status
    field "api_code (envelope code)", e.api_code
    field "api_message", e.api_message
    field "request_id", e.request_id
    field "attempts", e.attempts
    field "retry_after_seconds", e.retry_after_seconds

    e.api_errors.each do |detail|
      bullet "resource=#{detail.resource} code=#{detail.code || "-"} reason=#{detail.reason}"
    end

    # Sanitized diagnostics: safe to log, safe to attach to a support ticket.
    heading "  error.to_h — what you can safely log"
    json_block e.to_h

    serialized = e.to_json + e.full_message
    field "contains the API key", serialized.include?(ENV.fetch("PACTMAN_API_KEY"))

    # 404 is never retried, whatever the retry policy says.
    field "attempts made", "#{e.attempts} — not-found is not a transient failure"
  end

  # The bulk endpoint behaves differently, and this is the part that surprises
  # people: unmatched EINs come back on a successful 200 as item-level errors.
  # Only a request where *nothing* matched is a 404.
  heading "Bulk — mixed input returns HTTP 200, not an error"

  mixed = client.nonprofits.check_bulk([FixtureApi::EINS[:public_charity], no_record])

  field "status", mixed.status
  field "organizations returned", mixed.organizations.size
  field "not_found_eins", mixed.not_found_eins.join(", ")

  heading "Bulk — nothing matched at all"

  begin
    client.nonprofits.check_bulk([no_record])
    puts "  Unexpectedly succeeded."
  rescue NCP::ApiError => e
    field "class", e.class.name
    field "status", e.status
    field "api_message", e.api_message
  end
end

note "Distinguish \"we could not find it\" from \"we could not ask\". A 404 means the\n" \
     "record is absent; a timeout or a 503 means you learned nothing. Only the first\n" \
     "is a fact about the organization."
